File: backend/app/models/events.py
"""Pydantic models for the PyTorch analysis events.

Event-based state machine:
  1. Client connects via WebSocket
  2. Client sends code changes as they type (action: "update")
  3. When a `def` is detected → server sends func_pending event
  4. Server periodically runs lint and sends lint_feedback events
  5. Client sends "return" action → server triggers full analysis
  6. Server streams: architecture_detected → epoch_prediction(s) → training_summary
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class EventType(str, Enum):
    """Types of events streamed via WebSocket."""

    # Connection lifecycle
    CONNECTED = "connected"
    ERROR = "error"
    HEARTBEAT = "heartbeat"

    # Code editing phase
    FUNC_PENDING = "func_pending"
    LINT_FEEDBACK = "lint_feedback"

    # Analysis phase (triggered on "return")
    ANALYSIS_START = "analysis_start"
    ARCHITECTURE_DETECTED = "architecture_detected"
    EPOCH_PREDICTION = "epoch_prediction"
    TRAINING_SUMMARY = "training_summary"


class ClientAction(str, Enum):
    """Actions the client can send over the WebSocket."""

    UPDATE = "update"       # Code changed (typing)
    RETURN = "return"       # Submit for full analysis
    CANCEL = "cancel"       # Cancel ongoing analysis


class ClientMessage(BaseModel):
    """Message sent by the client over WebSocket."""

    action: ClientAction
    code: str = ""
    stream_interval: int = Field(
        default=30,
        ge=5,
        le=120,
        description="Interval in seconds between streamed epoch predictions",
    )


class LintIssue(BaseModel):
    """A single lint issue found in the code."""

    line: int
    column: int = 0
    severity: str = "warning"
    message: str


class PendingFunction(BaseModel):
    """A function definition detected as pending (not yet submitted)."""

    name: str
    line: int
    signature: str
    status: str = "pending"  # pending | analyzing | done


class ArchitectureInfo(BaseModel):
    """Detected PyTorch model architecture information."""

    model_type: str = ""
    layers: list[str] = Field(default_factory=list)
    optimizer: str = ""
    loss_function: str = ""
    dataset: str = ""
    batch_size: int | None = None
    learning_rate: float | None = None
    total_epochs: int | None = None
    device: str = "cpu"


class EpochMetrics(BaseModel):
    """Predicted metrics for a single epoch."""

    epoch: int
    total_epochs: int
    train_loss: float
    val_loss: float | None = None
    train_accuracy: float | None = None
    val_accuracy: float | None = None
    learning_rate: float | None = None
    elapsed_time_estimate: str = ""
    eta: str = ""


class TrainingSummary(BaseModel):
    """Summary of the entire predicted training run."""

    total_epochs: int
    final_train_loss: float
    final_val_loss: float | None = None
    final_train_accuracy: float | None = None
    final_val_accuracy: float | None = None
    best_epoch: int | None = None
    convergence_analysis: str = ""
    recommendations: list[str] = Field(default_factory=list)


class StreamEvent(BaseModel):
    """A single event sent over the WebSocket stream."""

    event_type: EventType
    data: dict[str, Any] = Field(default_factory=dict)
    message: str = ""
    timestamp: float = 0.0

File: frontend/src/types/events.ts
/** Shared types matching the FastAPI backend event models. */

export type EventType =
  | "connected"
  | "error"
  | "heartbeat"
  | "func_pending"
  | "lint_feedback"
  | "analysis_start"
  | "architecture_detected"
  | "epoch_prediction"
  | "training_summary";

export type ClientAction = "update" | "return" | "cancel";

export interface ClientMessage {
  action: ClientAction;
  code?: string;
  stream_interval?: number;
}

export interface LintIssue {
  line: number;
  column: number;
  severity: string;
  message: string;
}

export interface PendingFunction {
  name: string;
  line: number;
  signature: string;
  status: "pending" | "analyzing" | "done";
}

export interface ArchitectureInfo {
  model_type: string;
  layers: string[];
  optimizer: string;
  loss_function: string;
  dataset: string;
  batch_size: number | null;
  learning_rate: number | null;
  total_epochs: number | null;
  device: string;
}

export interface EpochMetrics {
  epoch: number;
  total_epochs: number;
  train_loss: number;
  val_loss: number | null;
  train_accuracy: number | null;
  val_accuracy: number | null;
  learning_rate: number | null;
  elapsed_time_estimate: string;
  eta: string;
}

export interface TrainingSummary {
  total_epochs: number;
  final_train_loss: number;
  final_val_loss: number | null;
  final_train_accuracy: number | null;
  final_val_accuracy: number | null;
  best_epoch: number | null;
  convergence_analysis: string;
  recommendations: string[];
}

export interface StreamEvent {
  event_type: EventType;
  data: Record<string, unknown>;
  message: string;
  timestamp: number;
}
